### PR TITLE
net/lorawan: Revert #16604 and fix NETOPT_RX_SYMBOL_TIMEOUT documentation

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
It turns out the SX127x DOES support 2 bytes symbol timeout. Therefore, #16604 makes no sense at all and the correct solution would be to change the NETOPT_RX_SYMBOL_TIMEOUT documentation to use `uint16_t` instead of `uint8_t`.

The worst thing is, #16604 also introduced a bug because the MSB was being written as the LSB. I guess I owe some beers...

This PR reverts such a disgrace and fixes the NETOPT documentation.

### Testing procedure

The Release tests are not enough to catch this bug. We should work on that.
The proper way to detect this bug is to follow a similar approach to https://github.com/RIOT-OS/RIOT/pull/16599#issuecomment-876300546.
I can provide a patch and test results.

### Issues/PRs references

#16604

### CODE REVIEW

**Comment by jia200x** [comment]
Test results:
Symbol time is ~32 ms in SF12. Therefore the timings are right.

```
2021-07-13 22:02:49,757 # main(): This is RIOT! (Version: 2021.10-devel-139-g61e24-pr/revert/rx_symbol)
2021-07-13 22:02:49,790 # [START] (10)
2021-07-13 22:02:50,119 # [STOP]: Took 326876 us
2021-07-13 22:02:50,121 # [START] (50)
2021-07-13 22:02:51,761 # [STOP]: Took 1632464 us
2021-07-13 22:02:51,763 # [2003620us]: RX_CONTINOUS
2021-07-13 22:02:51,768 # Initialization successful - starting the shell now
```

Using the patch:
```patch
diff --git a/tests/driver_sx127x/main.c b/tests/driver_sx127x/main.c
index 4fe1257c04..e36bec8e05 100644
--- a/tests/driver_sx127x/main.c
+++ b/tests/driver_sx127x/main.c
@@ -42,6 +42,8 @@
 
 #include "fmt.h"
 
+#include "mutex.h"
+
 #define SX127X_LORA_MSG_QUEUE   (16U)
 #define SX127X_STACKSIZE        (THREAD_STACKSIZE_DEFAULT)
 
@@ -53,6 +55,8 @@ static kernel_pid_t _recv_pid;
 static char message[32];
 static sx127x_t sx127x;
 
+static mutex_t lock;
+
 int lora_setup_cmd(int argc, char **argv)
 {
 
@@ -511,6 +515,10 @@ static void _event_cb(netdev_t *dev, netdev_event_t event)
             sx127x_set_sleep(&sx127x);
             break;
 
+        case NETDEV_EVENT_RX_TIMEOUT:
+            mutex_unlock(&lock);
+            break;
+
         default:
             printf("Unexpected netdev event received: %d\n", event);
             break;
@@ -518,6 +526,24 @@ static void _event_cb(netdev_t *dev, netdev_event_t event)
     }
 }
 
+static void _measure_rx_timeout(netdev_t *netdev, uint16_t rx_timeout)
+{
+    netdev->driver->set(netdev, NETOPT_RX_SYMBOL_TIMEOUT, &rx_timeout, sizeof(rx_timeout));
+    netopt_state_t state = NETOPT_STATE_IDLE;
+    netdev->driver->set(netdev, NETOPT_STATE, &state, sizeof(state));
+    uint32_t tmp = ztimer_now(ZTIMER_USEC);
+
+    if (rx_timeout) {
+        printf("[START] (%i)\n", rx_timeout);
+        mutex_lock(&lock);
+        printf("[STOP]: Took %li us\n", ztimer_now(ZTIMER_USEC) - tmp);
+    
+    }
+    else {
+        printf("[%"PRIu32"us]: RX_CONTINOUS\n", ztimer_now(ZTIMER_USEC));
+    }
+}
+
 void *_recv_thread(void *arg)
 {
     (void)arg;
@@ -545,6 +571,8 @@ int main(void)
     netdev_t *netdev = &sx127x.netdev;
 
     netdev->driver = &sx127x_driver;
+    mutex_init(&lock);
+    mutex_lock(&lock);
 
     if (netdev->driver->init(netdev) < 0) {
         puts("Failed to initialize SX127x device, exiting");
@@ -563,6 +591,15 @@ int main(void)
     }
 
     /* start the shell */
+    bool single = true;
+    netdev->driver->set(netdev, NETOPT_SINGLE_RECEIVE, &single, sizeof(single));
+
+    uint8_t sf = LORA_SF12;
+    netdev->driver->set(netdev, NETOPT_SPREADING_FACTOR, &sf, sizeof(sf));
+
+    _measure_rx_timeout(netdev, 10);
+    _measure_rx_timeout(netdev, 50);
+    _measure_rx_timeout(netdev, 0);
     puts("Initialization successful - starting the shell now");
     char line_buf[SHELL_DEFAULT_BUFSIZE];
```

**Comment by akshaim** [comment]
@jia200x 

Tested it on `b-l072z-lrwan1` and here are the results

```
2021-07-14 11:38:09,449 # main(): This is RIOT! (Version: 2021.10-devel-153-gb85e2-pr/revert/rx_symbol)
2021-07-14 11:38:09,481 # [START] (10)
2021-07-14 11:38:09,811 # [STOP]: Took 325743 us
2021-07-14 11:38:09,813 # [START] (50)
2021-07-14 11:38:11,452 # [STOP]: Took 1627057 us
2021-07-14 11:38:11,455 # [1997087us]: RX_CONTINOUS
2021-07-14 11:38:11,460 # Initialization successful - starting the shell now
```
The values seem to be a bit off for 10 waits it should be 327680, for 50 it should have been 1638400 for SF12 ( Symbol time of 32768 μs) but I think its okay.

Also what about `sx126x_netdev.c` . This change would break the current master for `sx126x_xx`

**Comment by jia200x** [comment]
> The values seem to be a bit off for 10 symbol waits; it should be 327680 μs. For 50, it should have been 1638400 μs for SF12 in 125kHz ( Symbol time of 32768 μs), but I think it's okay.

This is expected. There are prints in the middle and I'm relying on OS offload mechanisms for printing that.

> Also what about sx126x_netdev.c . This change would break the current master for sx126x_xx. Maybe a separate PR, I guess.

From what I see the `sx126x` is already using netdev. Why is it a breaking change?

**Comment by akshaim** [comment]
> `uint16_t`

I mean this https://github.com/RIOT-OS/RIOT/blob/master/drivers/sx126x/sx126x_netdev.c#L427-L430

**Comment by jia200x** [comment]
oh! I missed that one. I can amend it here

**Comment by jia200x** [comment]
This is the test output for `sx126x`:
```
> 2021-07-14 13:12:19,907 # main(): This is RIOT! (Version: 2021.10-devel-140-ga0e769-pr/revert/rx_symbol)
2021-07-14 13:12:19,923 # Initialization successful - starting the shell now
2021-07-14 13:12:19,925 # [START] (10)
2021-07-14 13:12:20,087 # [STOP] Took 162013 us
2021-07-14 13:12:20,088 # [START] (50)
2021-07-14 13:12:20,901 # [STOP] Took 811193 us
2021-07-14 13:12:20,903 # [1000635us]: RX_CONTINOUS
```

I used this patch (based on https://github.com/RIOT-OS/RIOT/pull/16599#issuecomment-876300546). The symbol time of SF11 and BW125 is 16ms. Therefore, T(10) ~160ms and T(50) = ~800 ms. The timings are right.

<details>

```patch
diff --git a/tests/driver_sx126x/Makefile b/tests/driver_sx126x/Makefile
index 68af3a584c..246dd6939c 100644
--- a/tests/driver_sx126x/Makefile
+++ b/tests/driver_sx126x/Makefile
@@ -6,5 +6,6 @@ USEMODULE += $(LORA_DRIVER)
 
 USEMODULE += shell
 USEMODULE += shell_commands
+USEMODULE += ztimer_usec
 
 include $(RIOTBASE)/Makefile.include
diff --git a/tests/driver_sx126x/main.c b/tests/driver_sx126x/main.c
index 104905e480..16587132eb 100644
--- a/tests/driver_sx126x/main.c
+++ b/tests/driver_sx126x/main.c
@@ -26,10 +26,12 @@
 #include "msg.h"
 #include "thread.h"
 #include "shell.h"
+#include "ztimer.h"
 
 #include "net/lora.h"
 #include "net/netdev.h"
 #include "net/netdev/lora.h"
+#include "mutex.h"
 
 #include "sx126x.h"
 #include "sx126x_params.h"
@@ -47,6 +49,8 @@ static char message[SX126X_MAX_PAYLOAD_LEN];
 
 static sx126x_t sx126x;
 
+static mutex_t lock;
+
 static void _event_cb(netdev_t *dev, netdev_event_t event)
 {
     if (event == NETDEV_EVENT_ISR) {
@@ -86,6 +90,10 @@ static void _event_cb(netdev_t *dev, netdev_event_t event)
             puts("Transmission timeout");
             break;
 
+        case NETDEV_EVENT_RX_TIMEOUT:
+            mutex_unlock(&lock);
+            break;
+
         default:
             printf("Unexpected netdev event received: %d\n", event);
             break;
@@ -318,6 +326,22 @@ static const shell_command_t shell_commands[] = {
     { NULL, NULL, NULL }
 };
 
+static void _measure_rx_timeout(netdev_t *netdev, uint16_t rx_timeout)
+{
+    netdev->driver->set(netdev, NETOPT_RX_SYMBOL_TIMEOUT, &rx_timeout, sizeof(rx_timeout));
+    netopt_state_t state = NETOPT_STATE_IDLE;
+    netdev->driver->set(netdev, NETOPT_STATE, &state, sizeof(state));
+    uint32_t t = ztimer_now(ZTIMER_USEC);
+    if (rx_timeout) {
+        printf("[START] (%i)\n", rx_timeout);
+        mutex_lock(&lock);
+        printf("[STOP] Took %li us\n", ztimer_now(ZTIMER_USEC) - t);
+    }
+    else {
+        printf("[%"PRIu32"us]: RX_CONTINOUS\n", ztimer_now(ZTIMER_USEC));
+    }
+}
+
 int main(void)
 {
     sx126x_setup(&sx126x, &sx126x_params[0], 0);
@@ -341,10 +365,19 @@ int main(void)
         return 1;
     }
 
+    mutex_init(&lock);
+    mutex_lock(&lock);
     /* start the shell */
     puts("Initialization successful - starting the shell now");
     char line_buf[SHELL_DEFAULT_BUFSIZE];
 
+    uint8_t sf = LORA_SF11;
+    netdev->driver->set(netdev, NETOPT_SPREADING_FACTOR, &sf, sizeof(uint8_t));
+
+    _measure_rx_timeout(netdev, 10);
+    _measure_rx_timeout(netdev, 50);
+    _measure_rx_timeout(netdev, 0);
+
     shell_run(shell_commands, line_buf, SHELL_DEFAULT_BUFSIZE);
     return 0;
 }

```

</details>

**Comment by akshaim** [comment]
@jia200x  Great, I think it's okay to squash now.

File: sys/include/net/netopt.h [hunk 1]
```diff
@@ -674,7 +674,7 @@ typedef enum {
     NETOPT_LORAWAN_MAX_RX_ERROR,
 
     /**
-     * @brief   (uint8_t) LoRaWAN maximum system overall timing error (symbols)
+     * @brief   (uint16_t) LoRaWAN maximum system overall timing error (symbols)
```
Comment (leandrolanzieri) [inline sys/include/net/netopt.h:677]: Wrong netopt?

Comment (jia200x) [reply sys/include/net/netopt.h:677]: yes... wrong netopt :/

Comment (leandrolanzieri) [reply sys/include/net/netopt.h:677]: Directly amend please

**Comment by jia200x** [comment]
> @jia200x Great, I think it's okay to squash now.

I would love to, but there's nothing to squash :P

**Review by leandrolanzieri — APPROVED** [review]
This properly reverts #16604, fixes the netopt documentation issue and fixes the netdev implementation of sx126x. @jia200x and @akshaim provided tests for this -> ACK

Everything but b299299ad3f10e45f202d8520fd19d8eab3c9ddc should be backported.

**Comment by jia200x** [comment]
thanks for the review!

**Comment by akshaim** [comment]
> > @jia200x Great, I think it's okay to squash now.
> 
> I would love to, but there's nothing to squash :P

Haha..  I thought sx126x commit was a fixup. Makes sense to have it as a separate commit as it helps with backporting :)

**Pull request merged**